File: dingtalk.py
#!/usr/bin/env python3
# 入乡随俗，使用下划线命名
import json
import re
import sys
from urllib.request import Request
from urllib.request import urlopen


def get_argument(name , defaultValue = None) :
    """
    获取参数
    :param name:
    :param defaultValue:
    :return:
    """
    indexOf = sys.argv.index(name)
    if indexOf + 1 < len(sys.argv) :
        return sys.argv[ indexOf + 1 ]

    return defaultValue


def get_hitokoto(api_url) :
    """
    获取一言
    :param api_url:
    :return:
    """
    return json.loads(urlopen(api_url).read().decode())


def post_with_json(robot_url , data) :
    """
    推送到钉钉
    :param robot_url:
    :param data:
    :return:
    """
    data = bytes(data , 'utf8')
    pushed = urlopen(Request(
        robot_url , headers = { 'Content-Type' : 'application/json;charset=utf-8' } , data = data
    ))

    return json.loads(pushed.read().decode('utf-8'))


def make_data(source , creator , sentence) :
    """
    组合数据
    :param source:
    :param creator:
    :param sentence:
    :return:
    """
    return json.dumps({
        'msgtype' : 'markdown' ,
        'markdown' : {
            'title' : '一言' ,
            'text' : "**%s** \n > %s「%s」" % (sentence , creator , source) ,
        }
    })


if __name__ == '__main__' :
    api_url = get_argument('--api-url')
    pattern = re.compile(r'^https?://')

    if pattern.match(api_url) is None :
        print('--api-url must be url')
        exit(127)

    robot_url = get_argument('--robot-url')
    if robot_url is None :
        print('--robot-url is required')
        exit(126)

    if pattern.match(robot_url) is None :
        print('--robot-url must be url')
        exit(125)

    hitokoto = get_hitokoto(api_url)
    response = post_with_json(
        robot_url , make_data(hitokoto[ 'from' ] , hitokoto[ 'creator' ] , hitokoto[ 'hitokoto' ])
    )

    if response[ 'errcode' ] != 0 :
        print(response[ 'errmsg' ])
        exit(response[ 'errcode' ])

    print('success')

File: dingtalk.js
#!/usr/bin/env node
const https = require('https')

function getArgument(name , defaultValue = null)
{
    const index = process.argv.indexOf(name)
    if (index === -1) {
        return defaultValue
    }
    return process.argv[index + 1]
}

function getHitokoto(apiUrl)
{
    return new Promise((resolve , reject) => {
        https.get(apiUrl , res => {
            res.on('data' , data => resolve(JSON.parse(data.toString())))
        }).on('error' , e => reject(e.message))
    })
}

function makeData(source , creator , sentence)
{
    return JSON.stringify({
        'msgtype': 'markdown' ,
        'markdown': {
            'title': '一言' ,
            'text': `**${sentence}** \n > ${creator}「${source}」` ,
        } ,
    })
}

async function postWithJson(url , data)
{
    const uriParams = url.split('/robot/')
    const options   = {
        path: `/robot/${uriParams[1]}` ,
        hostname: uriParams[0].replace('https://' , '') ,
        method: 'POST' ,
        headers: {
            'Content-Type': 'application/json;charset=utf-8' ,
        } ,
    }

    return new Promise((resolve , reject) => {
        const request = https.request(options , response => {
            response.on('data' , data => resolve(JSON.parse(data)))
        })

        request.on('error' , e => reject(e.message))
        request.write(data)
        request.end()
    })
}

(async () => {
    const robotUrl = getArgument('--robot-url')
    if (!/^https?:\/\//.test(robotUrl)) {
        return console.error('--robot-url must be url')
    }

    let apiUrl = getArgument('--api-url' , 'https://v1.hitokoto.cn/#')
    if (!/^https?:\/\//.test(apiUrl)) {
        return console.error('--api-url must be url')
    }

    // 获取一言
    const hitokoto = await getHitokoto(apiUrl ? apiUrl : null)

    // 推送到钉钉
    const response = await postWithJson(
        robotUrl ,
        makeData(hitokoto.from , hitokoto.creator , hitokoto.hitokoto) ,
    )

    if (response.errcode !== 0) {
        return console.error(response.errmsg)
    }

    console.log('success')
})()
